contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,320 @@
+
+contract RockPaperScissors is Ownable, Pausable {
+
+    using SafeMath for uint8;
+    using SafeMath for uint256;
+
+    mapping(address => uint256) public balances;
+    mapping(bytes32 => Game) public games;
+
+    address private constant NULL_ADDRESS = address(0);
+    string private constant INVALID_PLAYER_MSG = "Invalid player";
+    string private constant INVALID_MOVE_MSG = "Invalid move";
+    string private constant INVALID_STEP_MSG = "Invalid step";
+    string private constant GAME_NOT_EXPIRED_MSG = "Game has not expired";
+    string private constant SECRET_EMPTY_MSG = "Secret is empty";
+
+    uint256 constant FORFEIT_WINDOW = 1 days;
+    
+    enum Steps {
+        UNSTARTED, // 0
+        PLAYER_ONE_MOVED, // 1
+        PLAYER_TWO_MOVED, // 2
+        PLAYER_ONE_REVEALED // 3
+    }
+
+    enum GameResult {
+        DRAW, // 0
+        LEFTWIN, // 1
+        RIGHTWIN, // 2
+        INCORRECT // 3
+    }
+
+    enum Moves {
+        EMPTY, // 0
+        ROCK, // 1
+        PAPER, // 2
+        SCISSORS // 3
+    }
+
+    struct Game {
+        address playerOne;
+        Moves playerOneMove;
+        uint256 stake;
+        uint256 expiryDate;
+        Steps step;
+        address playerTwo;
+        bytes32 playerTwoMoveHash;
+    }
+
+    event PlayerOneMoves(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 stake,
+        uint256 amount
+    );
+
+    event PlayerTwoMoves(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount,
+        bytes32 moveHash,
+        uint256 expiryDate
+    );
+
+    event PlayerOneReveals(
+        bytes32 indexed gameKey,
+        address indexed player,
+        Moves move,
+        uint256 expiryDate
+    );
+
+    event PlayerTwoReveals(
+        bytes32 indexed gameKey,
+        address indexed player,
+        Moves move
+    );
+
+    event DrawRefund(
+        bytes32 indexed gameKey,
+        address indexed playerOne,
+        address indexed playerTwo,
+        uint256 playerOneRefund,
+        uint256 playerTwoRefund
+    );
+
+    event GamePayment(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount
+    );
+
+    event ForfeitPaid(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount
+    );
+
+    event WithDraw(
+        address indexed player,
+        uint amount
+    );
+
+    event PlayerOneEndsGame(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount
+    );
+
+    // The creator's move is embedded in the game key
+    function movePlayerOne(bytes32 gameKey, uint256 stake) external payable whenNotPaused {
+
+        require(gameKey != bytes32(0), "Game key hash is empty");
+        Game storage game = games[gameKey];
+        require(game.playerOne == NULL_ADDRESS, "Game key already used");
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
```

Comment:
I think you created a memory variable, so as to reuse it in the event.

Reply:
Yes. That had been the intention. Thanks for pointing out.